File: client/src/execution/logic/blocks.rs
//! The placeholder decode contract (oracle: `RunnableBlocks.scala`, pure half). The pipeline
//! emits `<div class="workbench" data-variants="<uri-encoded JSON>">`; the JSON is
//! `[{lang, source, viz?}]`. Languages are trimmed, blank-lang variants dropped, and an empty
//! list means the block is skipped. URI decoding is the view's job (it needs JS) — this stays
//! native-testable.

use serde::Deserialize;

#[derive(Debug, Clone, PartialEq, Eq, Deserialize)]
struct RawVariant {
    lang: String,
    source: String,
    #[serde(default)]
    viz: Option<String>,
}

/// One language rendition of a runnable block (oracle: shared `CodeVariant`).
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Variant {
    pub language: String,
    pub source: String,
}

/// Decode the (already URI-decoded) `data-variants` JSON. Malformed or empty → `None`
/// (the block is skipped, never a crash — authored content must not take the reader down).
pub fn parse_variants(json: &str) -> Option<Vec<Variant>> {
    let raw: Vec<RawVariant> = serde_json::from_str(json).ok()?;
    let variants: Vec<Variant> = raw
        .into_iter()
        .map(|v| Variant {
            language: v.lang.trim().to_owned(),
            source: v.source,
        })
        .filter(|v| !v.language.is_empty())
        .collect();
    if variants.is_empty() { None } else { Some(variants) }
}

/// Display name for a fence alias (oracle: `WorkbenchLogic.displayLang`).
pub fn display_lang(alias: &str) -> String {
    match alias.to_lowercase().as_str() {
        "cpp" | "c++" => "C++".to_owned(),
        "csharp" => "C#".to_owned(),
        other => {
            let mut chars = other.chars();
            match chars.next() {
                Some(first) => first.to_uppercase().collect::<String>() + chars.as_str(),
                None => String::new(),
            }
        }
    }
}

/// Seed the values grid from an authored case (oracle: `WorkbenchLogic.seedValues`).
pub fn seed_values(
    spec: &synapse_shared::execution::TestSpec,
    case_index: usize,
) -> std::collections::BTreeMap<String, String> {
    spec.cases
        .get(case_index)
        .map(|case| case.args.clone())
        .unwrap_or_default()
}

/// The active case's expected stdout, when declared.
pub fn expected_for(spec: &synapse_shared::execution::TestSpec, case_index: usize) -> Option<String> {
    spec.cases.get(case_index).and_then(|case| case.expected.clone())
}

#[cfg(test)]
#[allow(clippy::unwrap_used)]
mod tests {
    use super::*;

    #[test]
    fn decodes_single_and_adjacent_variants_in_order() {
        let json =
            r#"[{"lang":"python","source":"print(1)"},{"lang":"java","source":"class S {}","viz":"array"}]"#;
        let variants = parse_variants(json).unwrap();
        assert_eq!(variants.len(), 2);
        assert_eq!(variants[0].language, "python");
        assert_eq!(variants[1].source, "class S {}");
    }

    #[test]
    fn trims_langs_and_drops_blank_ones() {
        let json = r#"[{"lang":"  py  ","source":"a"},{"lang":"   ","source":"b"}]"#;
        let variants = parse_variants(json).unwrap();
        assert_eq!(variants.len(), 1);
        assert_eq!(variants[0].language, "py");
    }

    #[test]
    fn malformed_or_empty_means_skip() {
        assert_eq!(parse_variants("not json"), None);
        assert_eq!(parse_variants("[]"), None);
        assert_eq!(parse_variants(r#"[{"lang":" ","source":"x"}]"#), None);
    }

    #[test]
    fn display_names_read_well() {
        assert_eq!(display_lang("cpp"), "C++");
        assert_eq!(display_lang("python"), "Python");
        assert_eq!(display_lang("js"), "Js");
    }
}

File: web/src/lib/execution/blocks.ts
// The placeholder decode contract (oracle: client/src/execution/logic/blocks.rs, itself a port
// of `RunnableBlocks.scala`'s pure half). The markdown pipeline emits
// `<div class="workbench" data-variants="<uri-encoded JSON>">`; the JSON is
// `[{lang, source, viz?}]`. Languages are trimmed, blank-lang variants dropped, and an empty
// list means the block is skipped. URI decoding is the view/island's job (it needs the DOM) —
// this stays pure and native-testable.

import type { TestSpec } from "./judge";

interface RawVariant {
  lang: string;
  source: string;
  viz?: string | null;
}

/** One language rendition of a runnable block (oracle: shared `CodeVariant` + the
 *  positionally-paired `VizHints`). */
export interface Variant {
  language: string;
  source: string;
  /** The fence's `viz=<structure>[:<root>]` hint, raw (parsed on use). */
  viz: string | null;
}

/** Visualise needs a Python or Java variant with a `viz=` hint (oracle:
 *  `WorkbenchLogic.canVisualise`). */
export function canVisualise(variant: Variant): boolean {
  return variant.viz != null && ["python", "java"].includes(variant.language.toLowerCase());
}

/** Decode the (already URI-decoded) `data-variants` JSON. Malformed or empty → `null` (the block
 *  is skipped, never a crash — authored content must not take the reader down). */
export function parseVariants(json: string): Variant[] | null {
  let raw: RawVariant[];
  try {
    const parsed: unknown = JSON.parse(json);
    if (!Array.isArray(parsed)) return null;
    raw = parsed as RawVariant[];
  } catch {
    return null;
  }
  const variants: Variant[] = raw
    .map((v) => ({
      language: v.lang.trim(),
      source: v.source,
      viz: v.viz ?? null,
    }))
    .filter((v) => v.language !== "");
  return variants.length === 0 ? null : variants;
}

/** Display name for a fence alias (oracle: `WorkbenchLogic.displayLang`). */
export function displayLang(alias: string): string {
  switch (alias.toLowerCase()) {
    case "cpp":
    case "c++":
      return "C++";
    case "csharp":
      return "C#";
    case "rs":
      return "Rust";
    case "kt":
      return "Kotlin";
    case "js":
      return "JavaScript";
    case "ts":
      return "TypeScript";
    default: {
      const other = alias.toLowerCase();
      return other === "" ? "" : other[0].toUpperCase() + other.slice(1);
    }
  }
}

/** Seed the values grid from an authored case (oracle: `WorkbenchLogic.seedValues`). */
export function seedValues(spec: TestSpec, caseIndex: number): Record<string, string> {
  return spec.cases[caseIndex]?.args ?? {};
}

/** The active case's expected stdout, when declared. */
export function expectedFor(spec: TestSpec, caseIndex: number): string | null {
  return spec.cases[caseIndex]?.expected ?? null;
}

/**
 * Can a judged failure's input be reproduced in the VISIBLE tests panel? Only when every
 * declared arg has a value in the failure.
 *
 * A problem may be judged against a `<stem>.tests.json` sidecar the learner never sees, whose
 * arg ids need not match the authored fence's. Copying misaligned args would leave values under
 * keys with no input field, and `stdinFor` — which iterates the VISIBLE args — would then feed
 * the program something the judge never fed it. Extra keys are harmless: `stdinFor` ignores
 * them.
 */
export function canReproduce(spec: TestSpec, args: Record<string, string>): boolean {
  return spec.args.every((arg) => arg.id in args);
}
